Bad top-level program syntax results in assertion in `LexicalSortKey`

In the following test (in `main` branch), I'm expecting some diagnostics to be reported (parsing errors), but instead I get an assertion (stacktrace below).

From cursory investigation, the problem originates from `CreateSimpleProgram` (line 169: `nameLocation: new SourceLocation(firstGlobalStatement.GetFirstToken())`) when `GetFirstToken()` returns a default token.
Looking at a parsing test (below) we're producing a `GlobalStatement` with only missing nodes. 

```
        [Fact]
        public void TODO2()
        {
            var src = """
struct S { }
partial extension X for S { }
""";
            var comp = CreateCompilation(src);
            comp.VerifyDiagnostics(
                );
        }
```

```
        [Fact]
        public void TODO2()
        {
            var text = """ 
struct S { }
partial extension X for S { }
""";
            UsingTree(text,
                // (1,13): error CS1031: Type expected
                // struct S { }
                Diagnostic(ErrorCode.ERR_TypeExpected, "").WithLocation(1, 13),
                // (1,13): error CS1525: Invalid expression term 'partial'
                // struct S { }
                Diagnostic(ErrorCode.ERR_InvalidExprTerm, "").WithArguments("partial").WithLocation(1, 13),
                // (1,13): error CS1003: Syntax error, ',' expected
                // struct S { }
                Diagnostic(ErrorCode.ERR_SyntaxError, "").WithArguments(",").WithLocation(1, 13),
                // (2,1): error CS8803: Top-level statements must precede namespace and type declarations.
                // partial extension X for S { }
                Diagnostic(ErrorCode.ERR_TopLevelStatementAfterNamespaceOrType, "p").WithLocation(2, 1),
                // (2,29): error CS1002: ; expected
                // partial extension X for S { }
                Diagnostic(ErrorCode.ERR_SemicolonExpected, "}").WithLocation(2, 29),
                // (2,29): error CS1022: Type or namespace definition, or end-of-file expected
                // partial extension X for S { }
                Diagnostic(ErrorCode.ERR_EOFExpected, "}").WithLocation(2, 29)
                );

            N(SyntaxKind.CompilationUnit);
            {
                N(SyntaxKind.StructDeclaration);
                {
                    N(SyntaxKind.StructKeyword);
                    N(SyntaxKind.IdentifierToken, "S");
                    N(SyntaxKind.OpenBraceToken);
                    N(SyntaxKind.CloseBraceToken);
                }
                M(SyntaxKind.GlobalStatement);
                {
                    M(SyntaxKind.LocalDeclarationStatement);
                    {
                        M(SyntaxKind.VariableDeclaration);
                        {
                            M(SyntaxKind.IdentifierName);
                            {
                                M(SyntaxKind.IdentifierToken);
                            }
                            M(SyntaxKind.VariableDeclarator);
                            {
                                M(SyntaxKind.IdentifierToken);
                            }
                        }
                        M(SyntaxKind.SemicolonToken);
                    }
                }
                N(SyntaxKind.EndOfFileToken);
            }
            EOF();
        }
```

```
Message: 
System.InvalidOperationException : Assertion failed

  Stack Trace: 
ThrowingTraceListener.Fail(String message, String detailMessage) line 26
TraceInternal.Fail(String message)
Debug.Assert(Boolean condition)
LexicalSortKey.ctor(Int32 treeOrdinal, Int32 position) line 57
LexicalSortKey.ctor(SyntaxTree tree, Int32 position, CSharpCompilation compilation) line 63
LexicalSortKey.ctor(Location location, CSharpCompilation compilation) line 75
MergedTypeDeclaration.GetLexicalSortKey(CSharpCompilation compilation) line 149
SourceMemberContainerTypeSymbol.GetLexicalSortKey() line 959
LexicalOrderSymbolComparer.Compare(Symbol x, Symbol y) line 34
ImmutableArray`1.Sort(Int32 index, Int32 count, IComparer`1 comparer)
ImmutableArray`1.Sort(IComparer`1 comparer)
SourceNamespaceSymbol.GetMembers() line 156
SourceNamespaceSymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken) line 50
SourceModuleSymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken) line 271
SourceAssemblySymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken) line 924
CSharpCompilation.GetSourceDeclarationDiagnostics(SyntaxTree syntaxTree, Nullable`1 filterSpanWithinTree, Func`4 locationFilterOpt, CancellationToken cancellationToken) line 2967
CSharpCompilation.GetDiagnosticsWithoutFiltering(CompilationStage stage, Boolean includeEarlierStages, BindingDiagnosticBag builder, CancellationToken cancellationToken) line 2737
CSharpCompilation.GetDiagnostics(CompilationStage stage, Boolean includeEarlierStages, DiagnosticBag diagnostics, CancellationToken cancellationToken) line 2663
CSharpCompilation.GetDiagnostics(CompilationStage stage, Boolean includeEarlierStages, CancellationToken cancellationToken) line 2655
CSharpCompilation.GetDiagnostics(CancellationToken cancellationToken) line 2649
DiagnosticExtensions.VerifyDiagnostics[TCompilation](TCompilation c, DiagnosticDescription[] expected) line 108
RecordTests.TODO2() line 30421
```

I am getting errors in main for:
```
struct S { }
partial extension X for S { }
```

https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4GcFoMYIAEAyoQN6EC+AsAFAAOEyCUEANoQKYAeCnMWKAHsYhABqEAZkOQlyVIA==
```
error CS8803: Top-level statements must precede namespace and type declarations.
error CS1031: Type expected
error CS1525: Invalid expression term 'partial'
error CS1003: Syntax error, ',' expected
error CS1002: ; expected
error CS1022: Type or namespace definition, or end-of-file expected
```

--- a/src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs
+++ b/src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs
@@ -231,6 +231,15 @@ namespace Microsoft.CodeAnalysis.CSharp
 
         private static SingleNamespaceOrTypeDeclaration CreateSimpleProgram(GlobalStatementSyntax firstGlobalStatement, bool hasAwaitExpressions, bool isIterator, bool hasReturnWithExpression, ImmutableArray<Diagnostic> diagnostics)
         {
+            var nameLocation = new SourceLocation(firstGlobalStatement.GetFirstToken());
+
+            if (nameLocation.SourceTree is null)
+            {
+                nameLocation = new SourceLocation(firstGlobalStatement.GetFirstToken(includeSkipped: true));
+            }
+
+            Debug.Assert(nameLocation.SourceTree is not null);
+
             return new SingleTypeDeclaration(
                 kind: DeclarationKind.Class,
                 name: WellKnownMemberNames.TopLevelStatementsEntryPointTypeName,
@@ -241,7 +250,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                            (hasReturnWithExpression ? SingleTypeDeclaration.TypeDeclarationFlags.HasReturnWithExpression : SingleTypeDeclaration.TypeDeclarationFlags.None) |
                            SingleTypeDeclaration.TypeDeclarationFlags.IsSimpleProgram,
                 syntaxReference: firstGlobalStatement.SyntaxTree.GetReference(firstGlobalStatement.Parent),
-                nameLocation: new SourceLocation(firstGlobalStatement.GetFirstToken()),
+                nameLocation: nameLocation,
                 memberNames: s_emptyMemberNames,
                 children: ImmutableArray<SingleTypeDeclaration>.Empty,
                 diagnostics: diagnostics,

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs
@@ -9701,5 +9701,33 @@ public class C
             var conversionOutsideTopLevel = speculativeModelOutsideTopLevel.GetConversion(nodeToSpeculate.DescendantTokens().Single(n => n.ValueText == "x").Parent);
             Assert.Equal(ConversionKind.NoConversion, conversionOutsideTopLevel.Kind);
         }
+
+        [Fact]
+        [WorkItem("https://github.com/dotnet/roslyn/issues/67050")]
+        public void EmptyLocalDeclaration()
+        {
+            var src = """
+struct S { }
+partial ext X
+""";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (1,13): error CS1031: Type expected
+                // struct S { }
+                Diagnostic(ErrorCode.ERR_TypeExpected, "").WithLocation(1, 13),
+                // (1,13): error CS1525: Invalid expression term 'partial'
+                // struct S { }
+                Diagnostic(ErrorCode.ERR_InvalidExprTerm, "").WithArguments("partial").WithLocation(1, 13),
+                // (1,13): error CS1003: Syntax error, ',' expected
+                // struct S { }
+                Diagnostic(ErrorCode.ERR_SyntaxError, "").WithArguments(",").WithLocation(1, 13),
+                // (2,1): error CS8803: Top-level statements must precede namespace and type declarations.
+                // partial ext X
+                Diagnostic(ErrorCode.ERR_TopLevelStatementAfterNamespaceOrType, "").WithLocation(2, 1),
+                // (2,14): error CS1002: ; expected
+                // partial ext X
+                Diagnostic(ErrorCode.ERR_SemicolonExpected, "").WithLocation(2, 14)
+                );
+        }
     }
 }

--- a/src/Compilers/CSharp/Test/Syntax/Parsing/TopLevelStatementsParsingTests.cs
+++ b/src/Compilers/CSharp/Test/Syntax/Parsing/TopLevelStatementsParsingTests.cs
@@ -3537,5 +3537,63 @@ global using Bar x;
             }
             EOF();
         }
+
+        [Fact]
+        [WorkItem("https://github.com/dotnet/roslyn/issues/67050")]
+        public void EmptyLocalDeclaration()
+        {
+            var text = """ 
+struct S { }
+partial ext X
+""";
+            UsingTree(text,
+                // (1,13): error CS1031: Type expected
+                // struct S { }
+                Diagnostic(ErrorCode.ERR_TypeExpected, "").WithLocation(1, 13),
+                // (1,13): error CS1525: Invalid expression term 'partial'
+                // struct S { }
+                Diagnostic(ErrorCode.ERR_InvalidExprTerm, "").WithArguments("partial").WithLocation(1, 13),
+                // (1,13): error CS1003: Syntax error, ',' expected
+                // struct S { }
+                Diagnostic(ErrorCode.ERR_SyntaxError, "").WithArguments(",").WithLocation(1, 13),
+                // (2,1): error CS8803: Top-level statements must precede namespace and type declarations.
+                // partial ext X
+                Diagnostic(ErrorCode.ERR_TopLevelStatementAfterNamespaceOrType, "").WithLocation(2, 1),
+                // (2,14): error CS1002: ; expected
+                // partial ext X
+                Diagnostic(ErrorCode.ERR_SemicolonExpected, "").WithLocation(2, 14)
+                );
+
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.StructDeclaration);
+                {
+                    N(SyntaxKind.StructKeyword);
+                    N(SyntaxKind.IdentifierToken, "S");
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                M(SyntaxKind.GlobalStatement);
+                {
+                    M(SyntaxKind.LocalDeclarationStatement);
+                    {
+                        M(SyntaxKind.VariableDeclaration);
+                        {
+                            M(SyntaxKind.IdentifierName);
+                            {
+                                M(SyntaxKind.IdentifierToken);
+                            }
+                            M(SyntaxKind.VariableDeclarator);
+                            {
+                                M(SyntaxKind.IdentifierToken);
+                            }
+                        }
+                        M(SyntaxKind.SemicolonToken);
+                    }
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+            EOF();
+        }
     }
 }

